### PR TITLE
Explore ATH run against containerized selenium

### PR DESCRIPTION
**Pull request by olivergondza**
Permit ATH to utilize [selenium containers](https://github.com/SeleniumHQ/docker-selenium) to host browser, driver and selenium server. This si tangential to what ath-container do, but it integrates nicely with `buildPlugin`.

This is intended to be a way to use ATH from plugins source code.

Downstream PRs:
- https://github.com/jenkinsci/kerberos-sso-plugin/pull/19
- https://github.com/jenkinsci/openstack-cloud-plugin/pull/288
- https://github.com/jenkinsci/jms-messaging-plugin/pull/164

### CODE REVIEW

**Comment by olivergondza** [comment]
@jenkinsci/code-reviewers && @jglick, please have a look. Mainly at the documentation part: https://github.com/olivergondza/acceptance-test-harness/blob/prototype-containerized-selenium/docs/EXTERNAL.md (note the example repositories do not have the change merged jet)

File: src/main/java/org/jenkinsci/test/acceptance/FallbackConfig.java [hunk 1]
```diff
@@ -167,13 +180,63 @@ private WebDriver createWebDriver(TestName testName) throws IOException {
...
+                    } catch (IOException | InterruptedException e) {
+                        throw new Error("Failed removing container", e);
+                    }
+                }
+
+                @Override public String toString() {
+                    return "Kill and remove selenium container";
+                }
+            };
+            Thread.sleep(3000); // Give the container and selenium some time to spawn
```
Comment (amuniz) [inline src/main/java/org/jenkinsci/test/acceptance/FallbackConfig.java]: This seems a bit fragile. Isn't there a way to check if the container fully started? Maybe checking some logs entries?

Comment (vilacides) [reply src/main/java/org/jenkinsci/test/acceptance/FallbackConfig.java]: Another way is to poll for a 200 response for up to some time. That will save some time if it's faster than 3000

**Comment by jglick** [comment]
At one point @oleg-nenashev had been playing with related stuff.

**Comment by olivergondza** [comment]
Thanks for the reviews so far. I will proceed with integrating this by the end of the week provided there is no negative feedback.

**Comment by uhafner** [comment]
I tried to follow your example for the warnings plugin. I copied the tests to a new module and started them, but the tests do not start up due to an exception. I downloaded the kerberos plugin and here I also get the same exception: 
```
[ERROR] com.sonymobile.jenkins.plugins.kerberossso.KerberosSsoTest.explicitTicketAuth  Time elapsed: 4.938 s  <<< ERROR!
com.google.inject.ProvisionException: 
Unable to provision, see the following errors:

1) Error in custom provider, org.openqa.selenium.remote.UnreachableBrowserException: Could not start a new session. Possible causes are invalid address of the remote server or browser start-up failure.
Build info: version: '3.14.0', revision: 'aacccce0', time: '2018-08-02T20:19:58.91Z'
System info: host: 'Ullis-MBP-3.fritz.box', ip: '2003:cf:f32:d300:402b:fd9d:8528:49d0%en0', os.name: 'Mac OS X', os.arch: 'x86_64', os.version: '10.15.3', java.version: '1.8.0_242'
Driver info: driver.version: RemoteWebDriver
  at org.jenkinsci.test.acceptance.FallbackConfig.createWebDriver(FallbackConfig.java:293) (via modules: org.jenkinsci.test.acceptance.guice.World -> com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$OverrideModule -> org.jenkinsci.test.acceptance.FallbackConfig)
  at org.jenkinsci.test.acceptance.FallbackConfig.createWebDriver(FallbackConfig.java:293) (via modules: org.jenkinsci.test.acceptance.guice.World -> com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$OverrideModule -> org.jenkinsci.test.acceptance.FallbackConfig)
  while locating org.openqa.selenium.WebDriver
    for field at org.jenkinsci.test.acceptance.junit.DiagnosticRule.driver(DiagnosticRule.java:19)
  while locating org.jenkinsci.test.acceptance.junit.DiagnosticRule

1 error
Caused by: org.openqa.selenium.remote.UnreachableBrowserException: 
Could not start a new session. Possible causes are invalid address of the remote server or browser start-up failure.
Build info: version: '3.14.0', revision: 'aacccce0', time: '2018-08-02T20:19:58.91Z'
System info: host: 'Ullis-MBP-3.fritz.box', ip: '2003:cf:f32:d300:402b:fd9d:8528:49d0%en0', os.name: 'Mac OS X', os.arch: 'x86_64', os.version: '10.15.3', java.version: '1.8.0_242'
Driver info: driver.version: RemoteWebDriver
Caused by: java.net.ConnectException: Failed to connect to /127.0.0.1:64632
Caused by: java.net.ConnectException: Connection refused (Connection refused)

[INFO] 
[INFO] Results:
[INFO] 
[ERROR] Errors: 
[ERROR]   KerberosSsoTest.basicAuth » Provision Unable to provision, see the following e...
[ERROR]   KerberosSsoTest.explicitBasicAuth » Provision Unable to provision, see the fol...
[ERROR]   KerberosSsoTest.explicitTicketAuth » Provision Unable to provision, see the fo...
[ERROR]   KerberosSsoTest.kerberosTicket » Provision Unable to provision, see the follow...
[ERROR]   KerberosSsoTest.kerberosTicketWithBasicAuthEnabled » Provision Unable to provi...
[ERROR]   KerberosSsoTest.kerberosTicketWithBasicAuthEnabledJcasc » Provision Unable to ...
[INFO] 
[ERROR] Tests run: 6, Failures: 0, Errors: 6, Skipped: 0

```
Is macOS not supported for these kind of tests? ATH is running fine on macOS though.

**Comment by jglick** [comment]
> the browser inside container needs to be able to access Jenkins running on the host

FWIW I find it more straightforward to just run `mvn test`, and thus (with the default controller) Jenkins, inside a sibling container as well.

**Comment by olivergondza** [comment]
@jglick, wrapping Jenkins to a container (too) has a problem that it needs to reach to `MockUpdateCenter` (at least) running inside surefire process (on the host). Moving surefire into a container (the same or different from Jenkins) is something I have abandoned for its hideousness:

- https://github.com/jenkinsci/kerberos-sso-plugin/pull/19/commits/3644ab3c3837a4c754d41f3e0fb16d2cd9a9a5ae#diff-977da52725d56d537549658400297727R48-R67
- https://github.com/olivergondza/acceptance-test-harness/blob/6aafe6b4415cb06020418ef6c1f4088305df4a94/src/main/java/org/jenkinsci/test/acceptance/utils/ContainerizedRunner.java

So yeah, this is something to explore.

**Comment by uhafner** [comment]
I see. I can run the tests locally on macOS using the standard mechanism (e.g. JUT server), that works even from IntelliJ. During the CI build of my plugin I can switch to your new approach...

**Comment by olivergondza** [comment]
@uhafner, that might be the simplest way to go on Mac. Note just changing the variable from `<BROWSER>firefox-container</BROWSER>` to `<BROWSER>firefox</BROWSER>` should do the trick.

**Comment by jglick** [comment]
> wrapping Jenkins to a container (too) has a problem that it needs to reach to `MockUpdateCenter` (at least) running inside surefire process (on the host). Moving surefire into a container (the same or different from Jenkins)

FWIW I was not proposing any of these things. Rather, moving the entire `mvn test` inside a (`maven:3.6.3-jdk-8`, e.g.) container.

**Pull request merged**